File: AjaxVectorObjects/ClientScripts/VectorObjects/ShapeVObjectData.js
// Copyright (c) 2018 Aurigma Inc. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
//
Type.registerNamespace("Aurigma.GraphicsMill.AjaxControls.VectorObjects");

Aurigma.GraphicsMill.AjaxControls.VectorObjects.ShapeVObjectData = function (vObject) {
    Aurigma.GraphicsMill.AjaxControls.VectorObjects.ShapeVObjectData.initializeBase(this, [vObject]);

    if (vObject) {
        this.Pth = vObject.get_originalPath().toString();
        this.BW = vObject.get_borderWidth();
        this.BC = vObject.get_borderColor()._get_data();
        this.FC = vObject.get_fillColor()._get_data();
        this.FBW = vObject.get_fixedBorderWidth();
    }
};

Aurigma.GraphicsMill.AjaxControls.VectorObjects.ShapeVObjectData.applyState = function (shapeData, vObject) {
    if (vObject) {
        vObject.beginUpdate(); //call endUpdate in base class
        vObject.set_originalPath(new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.Path(shapeData.Pth));
        vObject.set_borderWidth(shapeData.BW);
        vObject.set_borderColor(Aurigma.GraphicsMill.AjaxControls.VectorObjects.Color._from_data(shapeData.BC));
        vObject.set_fillColor(Aurigma.GraphicsMill.AjaxControls.VectorObjects.Color._from_data(shapeData.FC));
        vObject.set_fixedBorderWidth(shapeData.FBW);
    }
    Aurigma.GraphicsMill.AjaxControls.VectorObjects.BaseRectangleVObjectData.applyState(shapeData, vObject);
};

Aurigma.GraphicsMill.AjaxControls.VectorObjects.ShapeVObjectData.registerClass("Aurigma.GraphicsMill.AjaxControls.VectorObjects.ShapeVObjectData", Aurigma.GraphicsMill.AjaxControls.VectorObjects.BaseRectangleVObjectData);

File: AjaxVectorObjects/VectorObjects/ShapeVObjectData.cs
// Copyright (c) 2018 Aurigma Inc. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
//
using Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math;
using System.Drawing;

namespace Aurigma.GraphicsMill.AjaxControls.VectorObjects
{
    public class ShapeVObjectData : BaseRectangleVObjectData
    {
        public ShapeVObjectData()
        {
        }

        public ShapeVObjectData(ShapeVObject obj)
            : base(obj)
        {
            Pth = obj.Path.ToSvgString();
            BC = obj.BorderColor;
            BW = obj.BorderWidth;
            FC = obj.FillColor;
            FBW = obj.FixedBorderWidth;
        }

        public override void ApplyState(VObject obj)
        {
            base.ApplyState(obj);

            var s = (ShapeVObject)obj;

            // Backward compatibility
            if (Pth == null)
            {
                var rectangle = new RectangleF(P[0].X, P[0].Y, P[1].X - P[0].X, P[1].Y - P[0].Y);
                s.Path = obj is EllipseVObject ? Path.CreateEllipsePath(rectangle) : Path.CreateRectanglePath(rectangle);
            }
            else
            {
                s.Path = Path.FromSvgString(Pth);
            }

            s.FillColor = FC;
            s.BorderColor = BC;
            s.BorderWidth = BW;
            s.FixedBorderWidth = FBW;
        }

        /// <summary>
        /// Path
        /// </summary>
        public string Pth { get; set; }

        /// <summary>
        /// BorderColor
        /// </summary>
        public Color BC { get; set; }

        /// <summary>
        /// FillColor
        /// </summary>
        public Color FC { get; set; }

        /// <summary>
        /// BorderWidth
        /// </summary>
        public float BW { get; set; }

        /// <summary>
        /// FixedBorderWidth
        /// </summary>
        public bool FBW { get; set; }
    }
}